patches/server/0970-New-specific-event-watching-command.patch

```diff
@@ -0,0 +1,220 @@
++import com.google.gson.JsonArray;
++import com.google.gson.JsonObject;
++import com.google.gson.internal.Streams;
++import com.google.gson.stream.JsonWriter;
++import io.papermc.paper.command.PaperSubcommand;
++import net.minecraft.server.MinecraftServer;
++import org.bukkit.command.CommandSender;
++import org.bukkit.event.Event;
++import org.bukkit.event.Listener;
++import org.checkerframework.checker.nullness.qual.NonNull;
++import org.checkerframework.checker.nullness.qual.Nullable;
++import org.checkerframework.framework.qual.DefaultQualifier;
++
++import java.io.IOException;
++import java.io.PrintStream;
++import java.io.StringWriter;
++import java.lang.reflect.Field;
++import java.nio.charset.StandardCharsets;
++import java.nio.file.Files;
++import java.nio.file.Path;
++import java.time.LocalDateTime;
++import java.time.format.DateTimeFormatter;
++import java.util.*;
++
++import static net.kyori.adventure.text.Component.text;
++import static net.kyori.adventure.text.format.NamedTextColor.GREEN;
++import static net.kyori.adventure.text.format.NamedTextColor.RED;
++
++@DefaultQualifier(NonNull.class)
++public class EventWatcherCommand implements PaperSubcommand {
++
++    /**
++     * The key is the equivalent of {@link Class#getSimpleName()} on the class,
++     * and value is the file we're trying to log to
++     */
++    private static final List<Watcher> watching = Collections.synchronizedList(new ArrayList<>()); //only sync in the rare case that multiple people are debugging, should be unnecessary
++    private static final DateTimeFormatter FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd_HH.mm.ss");
++
++
++    public static void tryLog(Event event, @Nullable Listener listener) { //listener should only be null on the initial event call
++        Optional<Watcher> watcher = watching.stream().filter(w -> w.eventPath.equals(event.getClass().getSimpleName())).findFirst();
++        watcher.ifPresent(value -> value.log(event, listener));
++    }
++
++    @Override
++    public boolean execute(CommandSender sender, String subCommand, String[] args) {
++        if(args.length < 1) {
++            sender.sendMessage(text("Invalid Usage! /paper watchevent [event path]", RED));
++            return true;
++        }
++        Optional<Watcher> watcherO = watching.stream().filter(w -> w.eventPath.equals(args[0])).findFirst();
++        if(watcherO.isPresent()) {
++            Watcher w = watcherO.get();
++            try {
++                w.dump();
++                sender.sendMessage(text("Successfully dumped event information to " + w.file, GREEN));
++            }catch(IOException ex) {
++                sender.sendMessage(text("Failed to dump event verbose! See the console for more info.", RED));
++                MinecraftServer.LOGGER.warn("Error occurred while creating event verbose files", ex);
++            }
++            watching.remove(w);
++        }else {
++            Path parent = Path.of("debug");
```

Comment:
Make this a static final field instead of creating it each time as it's always the same. Same goes for the other command.